nvmeof: expansion volume feature

gadididi opened the pull request:
# Describe what this PR does

Implements CSI `ControllerExpandVolume` and `NodeExpandVolume` for NVMe-oF volumes to support online volume expansion.

## Is there anything that requires special attention

**ControllerExpandVolume:**
- Calls directly to rbd `ControllerExpandVolume()` . 

**NodeExpandVolume:**
- Search for the by the current staging path the device path the pod is mounted to . 
- Resizes filesystem using `mount.NewResizeFs()` (supports ext4, xfs, btrfs)
- Skips filesystem resize for block mode volumes


FYI:
1. The NVMe-oF gateway supports namespace resizing via the ResizeNamespace GRPC call, which internally calls SPDK's `bdev_rbd_resize` to resize both the RBD image and the NVMeoF namespace. - BUT best practice is to resize directly the rbd itself. 

TODO:
 - add some unit tests 

## Related issues ##

https://github.com/ceph/ceph-csi/issues/5757


**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [ ] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [ ] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

gadididi committed:
nvmeof: add ControllerExpandVolume() implementation

ControllerExpandVolume() expand the nvmeof ns volume (and the rbd image size
)
Signed-off-by: gadi-didi <gadi.didi@ibm.com>

gadididi committed:
nvmeof: implement NodeExpandVolume for filesystem resize

Resize filesystem after ControllerExpandVolume expands the block device,
with support for old Kubernetes versions via staging path lookup
(as rbd driver does).

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

nixpanic commented:
> * The NVMe-oF gateway supports namespace resizing via the ResizeNamespace GRPC call, which internally calls SPDK's `bdev_rbd_resize` to resize both the RBD image and the NVMeoF namespace.
> 
> * If someone change externally the rbd image. automatically the namespace will be changed according to the rbd image. So maybe we should reconsider this approach: The NVMeoF's `ControllerExpandVolume()` will call to rbd's  `ControllerExpandVolume()`. that's it. for NodeExpandVolume, there is need to implement it, because the device is exposed as "nvme device" (and not "rbd device")

I do not have a preference for one way or the other. Maybe the gateway approach is more future proof, in case the gateway or spdk uses the size of the image at a later point in time?

nixpanic commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -330,16 +337,134 @@ func (cs *Server) ControllerModifyVolume(
+					return GetControllerPublishSecretRef(volumeId, driverType)
+				}
+		*/
+		secretName, secretNamespace, err := util.GetControllerPublishSecretRef(req.GetVolumeId(), util.RBDType)
```
The `ControllerExpandVolumeRequest` should contain secrets already, use those.

https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume

nixpanic replied:
you may remove the commented out code :smile:

gadididi replied:
@nixpanic , did you mean remove this entire section:
```go
	if secrets == nil {
		/*	Reuse the existing function - it's just getting Ceph credentials from k8s secret.
			GetControllerPublishSecretRef name is misleading here.
			another option is to create a new function like this one:

				func GetControllerSecretRef(volumeId, driverType string) (string, string, error) {
					return GetControllerPublishSecretRef(volumeId, driverType)
				}
		*/
		secretName, secretNamespace, err := util.GetControllerPublishSecretRef(req.GetVolumeId(), util.RBDType)
		if err != nil {
			log.ErrorLog(ctx, "controller expand secret not found: %v", err)

			return nil, status.Errorf(codes.Internal, "failed to get secrets: %v", err)
		}

		secrets, err = k8s.GetSecret(secretName, secretNamespace)
		if err != nil {
			return nil, status.Errorf(codes.Internal, "failed to get secrets: %v", err)
		}
	}

```
right?

nixpanic replied:
yes, correct. `secrets` should not be `nil` (in the near future).

nixpanic commented on internal/nvmeof/controller/controllerserver.go:
```diff
@@ -330,16 +337,134 @@ func (cs *Server) ControllerModifyVolume(
+	}
+
+	return &csi.ControllerExpandVolumeResponse{
+		CapacityBytes:         int64(namespaceSizeBytes),
```
does the gateway not round-up the size? The RBD-image has a certain block-size, so possibly the image a little larger than requested?

gadididi replied:
good catch I will check it. After size changes, I must return the current new size and not the requested size

gadididi replied:
update: nvmeof  `ControllerExpandVolume()` will call to rbd `ControllerExpandVolume()`

nixpanic commented on internal/nvmeof/nodeserver/nodeserver.go:
```diff
@@ -356,6 +357,78 @@ func (ns *NodeServer) NodeUnstageVolume(
+	}
+	defer ns.volumeLocks.Release(volumeID)
+
+	namespaceUUID, err := lookupNvmeVolumeMetadataStash(volumePath)
```
Is this the only thing where you need the metadata stash? Is there no reasonable syscall that we can use to identify the device by the `req.GetVolumePath()`?

If there is no need for the metadata stash, that would be quite nice.

gadididi replied:
I think I can retrieve the device path from `mount` e.g. `/dev/disk/by-id/nvme-uuid.12345678-1234-1234-1234-123456789abc`, 
then I can parse and get the UUID. 

but I was wondering, if the device is not mounted, it will fail. and metadata always succeed, what do you think?

Madhu-1 replied:
If the device is not mounted means that the NodeStage is not done, if its not done we need need to resize, we can return error until that time.

gadididi replied:
So, I am going to remove the stashing part, and just make small search:
something like this:

```Go
func getDeviceFromMount(mounter mount.Interface, mountPath string) (string, error) {
	// Get mount info
	mountInfos, err := mounter.List()
	if err != nil {
		return "", err
	}

	for _, mi := range mountInfos {
		if mi.Device == mountPath {
			// mi.Device is the device path
			// e.g., /dev/disk/by-id/nvme-uuid.12345678-1234-1234-1234-123456789abc
			return mi.Device, nil
		}
	}

	return "", fmt.Errorf("mount not found for path %s", mountPath)
}
```
when `mountPath` param is the `req.GetStagingTargetPath()`

Madhu-1 replied:
Will search be a costly operations when we have too many mounts in the system? if we will get into performance penalty lets stick to this one only. @nixpanic thoughts?

nixpanic replied:
IMHO stashing details is really ugly. There can be thousands of mounts on a container host, so the loop might take long. However, compared to the I/O done for resizing (metadata updates on the filesystem), this loop won't be significant. It is not like resizing is an operation that is done constantly.

My preference is to detect the device according to the path.

gadididi replied:
fixed

nixpanic commented on internal/nvmeof/nodeserver/nodeserver.go:
```diff
@@ -732,3 +805,40 @@ func getStagingTargetPath(req interface{}) string {
 }
+
+// getStagingPath returns the base staging path for a volume from its publish path.
+// For K8s < v1.19, NodeExpandVolume may only receive the publish path (where the pod sees the volume).
```
I assume this is mostly copied from the rbd nodeplugin. There is no support for Kuberneres 1.19 and older, there is no need to take that version into account anymore.

Madhu-1 replied:
Lets remove this one as well

Madhu-1 commented on internal/nvmeof/nodeserver/nodeserver.go:
```diff
@@ -332,11 +346,89 @@ func (ns *NodeServer) NodeUnstageVolume(
+			volumePath = req.GetVolumePath()
+			log.UsefulLog(ctx, "failed to find stagingTargetPath from volumePath %v: %v", volumePath, err)
+		}
+	}
```
Lets stick to the newer kubernetes supported version like > 1.30

Madhu-1 reviewed (commented):
small nits LGTM

gadididi committed:
nvmeof: add VolumeExpansion_ONLINE Capability

Add EXPAND_VOLUME capability to NodeServer for online volume expansion support

Signed-off-by: gadi-didi <gadi.didi@ibm.com>

nixpanic commented on internal/nvmeof/nvmeof.go:
```diff
@@ -205,6 +205,35 @@ func (gw *GatewayRpcClient) SetQoSLimitsForNamespace(
 }
 
+// ResizeNamespace resizes a NVMeoF namespace to a new size.
+func (gw *GatewayRpcClient) ResizeNamespace(
```
I think you can drop this commit now too.

gadididi replied:
removed

nixpanic reviewed (commented):
looks good to me, just need to drop the 1st commit as it is unused?

Madhu-1 commented on internal/nvmeof/nvmeof.go:
```diff
@@ -238,6 +238,38 @@ func (gw *GatewayRpcClient) GetUUIDBySubsystemAndNameSpaceID(
 }
 
+// GetNamespaceSize gets the size of a namespace in bytes.
+func (gw *GatewayRpcClient) GetNamespaceSize(
```
Lets remove it as its not used anywhere?

gadididi replied:
removed
